devito/finite_differences/finite_difference.py

```diff
@@ -229,32 +236,45 @@ def generic_derivative(expr, dim, fd_order, deriv_order, stagger=None, symbolic=
         Derivative order, e.g. 2 for a second-order derivative.
     stagger : Side, optional
         Shift of the finite-difference approximation.
+    x0 : dict, optional
+        Origin of the finite-difference scheme as a map dim: origin_dim.
 
     Returns
     -------
     expr-like
         ``deriv-order`` derivative of ``expr``.
     """
-    diff = dim.spacing
-
+    # First order derivative with 2nd order FD is highly non-recommended so taking
+    # first order fd that is a lot better
+    if deriv_order == 1 and fd_order == 2 and not symbolic:
+        fd_order = 1
     # Stencil positions
-    indices, x0 = generate_indices(expr, dim, diff, fd_order, stagger=stagger)
+    indices, x0 = generate_indices(expr, dim, fd_order, x0=x0)
 
     # Finite difference weights from Taylor approximation with these positions
     if symbolic:
         c = symbolic_weights(expr, deriv_order, indices, x0)
     else:
         c = finite_diff_weights(deriv_order, indices, x0)[-1][-1]
 
-    # Loop through positions
-    deriv = 0
-    all_dims = tuple(set((dim,) + tuple(i for i in expr.indices if i.root == dim)))
-    for i in range(len(indices)):
-        subs = dict((d, indices[i].subs({dim: d, diff: matvec.val*diff}))
-                    for d in all_dims)
-        deriv += expr.subs(subs) * c[i]
+    return indices_weights_to_fd(expr, dim, indices, c, matvec=matvec.val)
 
-    # Evaluate up to _PRECISION digits
-    deriv = deriv.evalf(_PRECISION)
 
-    return deriv
+def indices_weights_to_fd(expr, dim, inds, weights, matvec=1):
+    """Expression from lists of indices and weights."""
+    diff = dim.spacing
+    deriv = 0
+    all_dims = tuple(set((expr.index(dim),) +
```

Comment:
potential BS:

isnt this just `.index_ref[dim]`? instead of `.index(dim)`

Reply:
No it's not. `index_ref` is the map `self.dimensions : self.indices` while `index(dim)` takes any Dimension as input and check wether it is in the indices or if its root is in it. Two completely different things